edb/edgeql/compiler/normalization.py

```diff
@@ -0,0 +1,539 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2020-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either nodeess or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+"""EdgeQL expression normalization functions."""
+
+
+from __future__ import annotations
+from typing import *
+
+import copy
+import functools
+
+from edb.common.ast import base
+from edb.edgeql import ast as qlast
+from edb.schema import schema as s_schema
+from edb.schema import name as sn
+
+
+@functools.singledispatch
+def normalize(
+    node: Any,
+    *,
+    schema: s_schema.Schema,
+    modaliases: Mapping[Optional[str], str],
+    ignorenames: Set[str] = frozenset(),
+) -> Any:
+    raise ValueError(f'normalize: cannot handle {node!r}')
+
+
+def _normalize_generic_field(
+    node: qlast.Base,
+    field: str,
+    value: Any,
+    *,
+    schema: s_schema.Schema,
+    modaliases: Mapping[Optional[str], str],
+    ignorenames: Set[str] = frozenset(),
+) -> None:
+    if isinstance(value, qlast.Base):
+        setattr(
+            node,
+            field,
+            normalize(
+                value,
+                schema=schema,
+                modaliases=modaliases,
+                ignorenames=ignorenames,
+            )
+        )
+    elif isinstance(value, list):
+        if value and isinstance(value[0], qlast.Base):
+            setattr(
+                node,
+                field,
+                [
+                    normalize(
+                        el,
+                        schema=schema,
+                        modaliases=modaliases,
+                        ignorenames=ignorenames,
+                    ) for el in value
+                ]
+            )
+
```

Comment:
```
else:
    raise AssertionError
```

Reply:
Actually this was deliberately omitted, in this generic field normalization we recursively process other AST nodes and lists of them, but we ignore other types that are assumed to be leaves in this recursion (and handled by the AST-node-specific handler).

I've renamed the function and added a comment to address this.